### PR TITLE
Plumb dart timeline events to mojo:tracing.

### PR DESCRIPTION
**Pull request by ppiet**
This patch:

 - makes dart traces (coming from the VM and the timeline API)
   available to mojo:tracing
 - adds --complete-timeline parameter, which makes the Dart VM enable all
   categories of traces from the beginning, allowing to do startup tracing

This is based on the equivalent support in dart content handler. In particular,
dart_tracing.{h,cc} are imported with only mechanical changes (includes and
include guard).

### CODE REVIEW

**Comment by ppiet** [comment]
@chinmaygarde @johnmccutchan - ptal

File: sky/engine/core/script/dart_init.cc [hunk 1]
```diff
@@ -102,6 +102,10 @@ static const char* kDartStartPausedArgs[]{
     "--pause_isolates_on_start",
 };
 
+static const char* kCompleteTimelineArgs[]{
+    "--complete-timeline",
```
Comment (krisgiesing) [inline sky/engine/core/script/dart_init.cc]: The other flags use underscore conventions here. Should this be `--complete_timeline`?

**Comment by johnmccutchan** [comment]
I do not recommend using --complete-timeline for startup tracing.

**Comment by ppiet** [comment]
@johnmccutchan what do you recommend instead? Without this option we don't get the timeline traces. (neither from flutter w/ this patch nor from the dart content handler)

File: sky/engine/core/script/dart_init.cc (hunk 1 above)
Comment (ppiet) [inline sky/engine/core/script/dart_init.cc]: Done.

**Comment by ppiet** [comment]
Can we merge this? As per https://codereview.chromium.org/1893223004/ and offline discussion with @johnmccutchan , we will need a new flag in the dart VM to make this work without `--complete-timeline`. In the meantime, this patch works and is useful (@emembrives has been patching it in locally).

**Comment by eseidelGoogle** [comment]
@abarth is probably your correct reviewer, he's OOO until Weds.

**Comment by johnmccutchan** [comment]
@ppiet Why can't the Dart traces be fetched over the service protocol directly?

**Comment by ppiet** [comment]
@johnmccutchan what is "service protocol"? We need the dart traces to be available to mojo:tracing, so that one can take a unified trace representing the events in dart content handler, flutter content handler, and other mojo apps.

**Comment by ppiet** [comment]
@abarth ptal. @johnmccutchan can you explain the point about the service protocol above?

File: sky/shell/platform/mojo/dart_tracing.cc [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+    Dart_GlobalTimelineSetRecordedStreams(DART_TIMELINE_STREAM_DART);
+  } else {
+    // TODO(johnmccutchan): Respect |categories|.
+    EnableAll();
+  }
+}
+
+void DartTimelineController::EnableAll() {
+  Dart_GlobalTimelineSetRecordedStreams(DART_TIMELINE_STREAM_DART);
+  Dart_GlobalTimelineSetRecordedStreams(DART_TIMELINE_STREAM_ALL |
```
Comment (chinmaygarde) [inline sky/shell/platform/mojo/dart_tracing.cc]: I believe you can replace both these lines with just `Dart_GlobalTimelineSetRecordedStreams(DART_TIMELINE_STREAM_ALL)`

**Comment by johnmccutchan** [comment]
@ppiet The vm service protocol is a JSON RPC protocol for communicating with the VM. It is what Observatory uses.

**Comment by chinmaygarde** [comment]
I believe what @johnmccutchan is referring to is connecting to the Dart observatory using a WebSocket connection and make specific requests to start, stop and collect tracing information. Something like https://github.com/flutter/flutter/pull/3393/files.

lgtm as a starting point for getting immediately useful tracing information.

File: sky/shell/platform/mojo/dart_tracing.cc [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+  size_t end = 0;
+  int curly_braces = 0;
+  for (size_t i = 0; i < length; i++) {
+    if (data[i] == '{') {
+      if ((curly_braces == 0) && (start == kInvalidIndex)) {
+        start = i;
+      }
+      curly_braces++;
+    }
+    if (data[i] == '}') {
```
Comment (abarth) [inline sky/shell/platform/mojo/dart_tracing.cc]: What if these curly braces are inside string literals?

File: sky/engine/public/platform/sky_settings.h [hunk 4]
```diff
@@ -17,6 +17,7 @@ struct SkySettings {
   bool start_paused = false;
   bool enable_dart_checked_mode = false;
   bool trace_startup = false;
+  bool complete_timeline = false;
```
Comment (abarth) [inline sky/engine/public/platform/sky_settings.h]: What is "complete timeline" ?  We should have a better name than this.  Why would I ever want an incomplete timeline?

Comment (krisgiesing) [inline sky/engine/public/platform/sky_settings.h]: I believe this nomenclature comes from a Dart VM flag.

Comment (abarth) [inline sky/engine/public/platform/sky_settings.h]: Yeah, but the name doesn't make sense outside that context.  (I'm not even sure what the name means in the context of the Dart VM.)

Comment (johnmccutchan) [inline sky/engine/public/platform/sky_settings.h]: By default the VM timeline uses a ring buffer and old events are dropped.

--complete_timeline is short hand for:

--timeline_streams=all
--timeline_recorder=endless

All timeline categories are enabled and we use the 'endless' recorder which will keep allocating memory for timeline events until malloc fails.

--startup_timeline is short hand for:

--timeline_streams=all
--timeline_recorder=startup

All timeline categories are enabled and we use the 'startup' recorder which will record up to 8k events and then start dropping events.

Comment (abarth) [inline sky/engine/public/platform/sky_settings.h]: That makes sense.  At this layer of the system, we'll probably need a different name given that we'll still use a ring buffer for the `base` trace events.

File: sky/shell/platform/mojo/dart_tracing.cc (hunk 3 above)
Comment (ppiet) [inline sky/shell/platform/mojo/dart_tracing.cc]: Done - switched to use proper JSON parser from base.

File: sky/engine/public/platform/sky_settings.h (hunk 4 above)
Comment (ppiet) [inline sky/engine/public/platform/sky_settings.h]: Renamed to "vm-complete-timeline". (BTW, one could argue that that `base` trace events are not "timeline" events, so maybe there was no conflict in the first place)

File: sky/shell/platform/mojo/dart_tracing.cc (hunk 2 above)
Comment (ppiet) [inline sky/shell/platform/mojo/dart_tracing.cc]: Done.

**Comment by ppiet** [comment]
@chinmaygarde @johnmccutchan thanks for clarification regarding the vm service protocol. How would you suggest that we integrate this with mojo:tracing? For tracing mojo apps it is essential that the tracing information is provided to mojo:tracing, so that one can take a unified trace covering all apps that are running.

@abarth thanks, I addressed comments so far, ptal.

**Comment by johnmccutchan** [comment]
@ppiet I think that @chinmaygarde explained it. You can fetch the trace information over the service protocol instead of gathering it this way.

**Comment by ppiet** [comment]
@johnmccutchan I must be still missing something, because connecting to flutter observatory over websocket (from where?) doesn't seem to be a solution that ends up making mojo calls to the tracing service.

**Comment by ppiet** [comment]
Rebased the PR on upstream master. @abarth @chinmaygarde should I squash the commits?

**Comment by chinmaygarde** [comment]
Yes please. Thank you.

**Comment by ppiet** [comment]
Done.

**Comment by Hixie** [comment]
What is the status of this PR?

**Comment by ppiet** [comment]
@Hixie good to merge as far as I am concerned.

**Comment by Hixie** [comment]
Ok, landing it. Thanks!

**Pull request merged**